lib/posix/pthread_key.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <kernel.h>
+#include <misc/dlist.h>
+#include <semaphore.h>
+#include <errno.h>
+#include <posix/pthread.h>
+
+static sem_t sem;
+
+/**
+ * @brief Create a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_create(pthread_key_t *key,
+		void (*destructor)(void *))
+{
+	pthread_key_obj *newKey;
+
+	*key = NULL;
+
+	sem_init(&sem, 0, 1);
+
+	newKey = (pthread_key_obj *)k_malloc(sizeof(pthread_key_obj));
+
+	if (newKey == NULL) {
+		errno = ENOMEM;
+		return -1;
```

Comment:
This function does not set `errno`, but rather return the error code directly:

```
RETURN VALUE
       If  successful,  the pthread_key_create() function shall store the newly created key value
       at *key and shall return zero. Otherwise, an error number shall be  returned  to  indicate
       the error.

```

Reply:
Yes, you are right. I will change this accordingly to return the error codes directly in other APIs as well